dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/loadbalance/AbstractLoadBalance.java

```diff
@@ -75,10 +75,16 @@ protected int getWeight(Invoker<?> invoker, Invocation invocation) {
         if (weight > 0) {
             long timestamp = invoker.getUrl().getParameter(REMOTE_TIMESTAMP_KEY, 0L);
             if (timestamp > 0L) {
-                int uptime = (int) (System.currentTimeMillis() - timestamp);
+                long uptime = System.currentTimeMillis() - timestamp;

```

Comment:
should we change this method to

```java
    protected int getWeight(Invoker<?> invoker, Invocation invocation) {
        int weight = invoker.getUrl().getMethodParameter(invocation.getMethodName(), WEIGHT_KEY, DEFAULT_WEIGHT);
        if (weight > 0) {
            long timestamp = invoker.getUrl().getParameter(TIMESTAMP_KEY, 0L);
            if (timestamp > 0L) {
                int uptime = (int) Math.max(System.currentTimeMillis() - timestamp, 0);
                int warmup = invoker.getUrl().getParameter(WARMUP_KEY, DEFAULT_WARMUP);
                if (uptime >= 0 && uptime < warmup) {
                    weight = calculateWarmupWeight(uptime, warmup, weight);
                }
            }
        }
        return weight >= 0 ? weight : 0;
    }
```

Reply:
the diff is

```diff
index ed6f3c4e9..14640d4fb 100644
--- a/dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/loadbalance/AbstractLoadBalance.java
+++ b/dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/loadbalance/AbstractLoadBalance.java
@@ -24,6 +24,7 @@ import org.apache.dubbo.rpc.cluster.LoadBalance;
 
 import java.util.List;
 
+import static org.apache.dubbo.common.constants.CommonConstants.TIMESTAMP_KEY;^M
 import static org.apache.dubbo.rpc.cluster.Constants.DEFAULT_WARMUP;
 import static org.apache.dubbo.rpc.cluster.Constants.DEFAULT_WEIGHT;
 import static org.apache.dubbo.rpc.cluster.Constants.REMOTE_TIMESTAMP_KEY;
@@ -73,11 +74,11 @@ public abstract class AbstractLoadBalance implements LoadBalance {
     protected int getWeight(Invoker<?> invoker, Invocation invocation) {
         int weight = invoker.getUrl().getMethodParameter(invocation.getMethodName(), WEIGHT_KEY, DEFAULT_WEIGHT);
         if (weight > 0) {
-            long timestamp = invoker.getUrl().getParameter(REMOTE_TIMESTAMP_KEY, 0L);
+            long timestamp = invoker.getUrl().getParameter(TIMESTAMP_KEY, 0L);^M
             if (timestamp > 0L) {
-                int uptime = (int) (System.currentTimeMillis() - timestamp);
+                int uptime = (int) Math.max(System.currentTimeMillis() - timestamp, 0);^M
                 int warmup = invoker.getUrl().getParameter(WARMUP_KEY, DEFAULT_WARMUP);
-                if (uptime > 0 && uptime < warmup) {
+                if (uptime >= 0 && uptime < warmup) {^M
                     weight = calculateWarmupWeight(uptime, warmup, weight);
                 }
             }
```

Reply:
I don't think so,  if the provider is started 24 days ago, then `System.currentTimeMillis() - timestamp` will be bigger than `Integer.MAX_VALUE`.

Consider that `System.currentTimeMillis() = 1566209746000`(2019-08-19 18:15:46), and `timestamp = 1561914711000`(2019-07-01 01:11:51),  then `int uptime = (int) Math.max(System.currentTimeMillis() - timestamp, 0)` is `67704`, this will result in a smaller weight.